### DISCUSSION TITLE
DataCoord is not ready, error=service not ready: Abnormal

### DISCUSSION THREAD

**Discussion by kietheros** [milvus-io/milvus#32459] [Q&A and General discussion]
My deployment:
- Milvus version: v2.3.12
- Deployment mode(standalone or cluster): cluster
- MQ type(rocksmq, pulsar or kafka): kafka 
- SDK version(e.g. pymilvus v2.0.0rc2): pymilvus 2.2.0
- OS(Ubuntu or CentOS): Ubuntu 22.04
- CPU/Memory: Intel Core i9-9900K

During operation, I encountered several warning logs indicating issues with the DataCoord service. Initially, there were warning logs related to session revocation failures and context deadline exceeded errors. 
```
milvus-datacoord  | [2024/04/17 16:11:45.576 +00:00] [WARN] [sessionutil/session_util.go:962] ["failed to revoke session"] [sessionKey=] [error="context deadline exceeded"]
milvus-datacoord  | [2024/04/17 16:11:45.576 +00:00] [INFO] [sessionutil/session_util.go:964] ["revoke session successfully"] [sessionKey=]
milvus-datacoord  | {"level":"warn","ts":"2024-04-17T16:11:45.576Z","logger":"etcd-client","caller":"v3@v3.5.5/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0013e01c0/milvus-etcd:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
milvus-datacoord  | [2024/04/17 16:11:48.900 +00:00] [WARN] [sessionutil/session_util.go:551] ["fail to retry keepAliveOnce"] [serverName=indexcoord] [LeaseID=7587878065739745264] [error="attempt #0: etcdserver: requested lease not found: attempt #1: context deadline exceeded: attempt #2: etcdserver: requested lease not found"]
```

After that, rootcoord, querycoord, datacoord,... restarts, all services run but datacoord outputs many errors

```
datacoord  | [2024/04/19 05:24:05.221 +00:00] [WARN] [datacoord/index_service.go:544] ["DataCoord 68 is not ready"] [traceID=7898ceade2ab03d9f1b8f9f79136e976] [collectionID=448554834316755802] [indexName=] [error="service not ready[datacoord=68]: Abnormal"]
datacoord  | [2024/04/19 05:24:15.222 +00:00] [INFO] [datacoord/index_service.go:539] ["receive DescribeIndex request"] [traceID=7898ceade2ab03d9f1b8f9f79136e976] [collectionID=448554834316755802] [indexName=] [timestamp=0]
```

Here are all log files at the error time
[datacoord.log](https://github.com/milvus-io/milvus/files/15033991/datacoord.log)
[querycoord.log](https://github.com/milvus-io/milvus/files/15033993/querycoord.log)
[datanode.log](https://github.com/milvus-io/milvus/files/15034001/datanode.log)
[indexnode.log](https://github.com/milvus-io/milvus/files/15034002/indexnode.log)

**Comment by yhmo** [comment]
Similar to this issue: https://github.com/milvus-io/milvus/issues/31858
The etcd ran into trouble. Possible reason: etcd is running on a poor performance disk.

**Reply by kietheros** [reply]
Thank you for your reply. I will move the etcd data directory to an NVMe disk and check if this error occurs again

**Comment by kietheros** [comment]
@yhmo Hi, I encountered this error again.
- DataCoord 157 is not ready
- ["register active key to etcd failed"] [error="etcdserver: requested lease not found"]

Could you please review these logs? Currently, I am using SSD disk. The system load is low, insert ~ 3 records/s and a query ~ 5 queries/s. I believe my etcd cluster performance is adequate to handle the requests.

Here are the etcd benchmark:
```
benchmark --endpoints=node-1:2379,node-2:2379,node-3:2379 --conns=100 --clients=1000     put --key-size=8 --sequential-keys --total=100000 --val-size=25
Summary:
  Total:	2.3008 secs.
  Slowest:	0.0666 secs.
  Fastest:	0.0063 secs.
  Average:	0.0226 secs.
  Stddev:	0.0090 secs.
  Requests/sec:	43464.0877
  ```
milvus.py
```
  session:
    ttl: 20 # ttl value when session granting a lease to register service
    retryTimes: 30 # retry times when session sending etcd requests
```
When I restart the cluster, Milvus works again, so I think there is a problem with the services.

[datacoord.log](https://github.com/milvus-io/milvus/files/15372282/datacoord.log)
[etcd.log](https://github.com/milvus-io/milvus/files/15372283/etcd.log)

**Comment by xiaofan-luan** [comment]
watch timeout: "timeout and stop timer: wait for channel ACK timeout"

you need to offer log from datanode and check why channel watch is timeout

**Reply by kietheros** [reply]
@xiaofan-luan Thank you for your reply! I found that the RAM usage of one of my servers was full during certain periods of time. Kafka and Minio may not have been working properly. However, after fixing Kafka and Minio, I had to force restart the Milvus cluster to get it working again.
